File: week2/secp256k1/python/affine.py
from typing import Optional, Tuple

# secp256k1 field modulus
P = 2**256 - 2**32 - 977

def inv(a: int) -> int:
    """Modular inverse using Fermat's Little Theorem: a⁻¹ ≡ a^(p−2) mod p"""
    return pow(a, P - 2, P)

def affine_add(
    p1: Optional[Tuple[int, int]],
    p2: Optional[Tuple[int, int]]
) -> Optional[Tuple[int, int]]:
    """
    Add two points in affine coordinates on the secp256k1 elliptic curve.
    Returns None for the point at infinity.
    """
    if p1 is None:
        return p2
    if p2 is None:
        return p1

    x1, y1 = p1
    x2, y2 = p2

    # P + (-P) = ∞
    if x1 == x2 and (y1 + y2) % P == 0:
        return None

    if x1 != x2:
        # λ = (y₂ - y₁) / (x₂ - x₁)
        lam = (y2 - y1) * inv(x2 - x1) % P
    else:
        # Point doubling: λ = (3x₁²) / (2y₁)
        lam = (3 * x1 * x1) * inv(2 * y1) % P

    # x₃ = λ² − x₁ − x₂
    xr = (lam * lam - x1 - x2) % P
    # y₃ = λ(x₁ − x₃) − y₁
    yr = (lam * (x1 - xr) - y1) % P

    return (xr, yr)

File: week2/secp256k1/rust/src/affine.rs
use num_bigint::BigUint;
use num_traits::{One, Zero};
use once_cell::sync::Lazy;

// Field modulus p = 2^256 - 2^32 - 977
pub static P: Lazy<BigUint> =
    Lazy::new(|| (BigUint::one() << 256) - (BigUint::one() << 32) - BigUint::from(977u32));

pub fn modp() -> &'static BigUint {
    &*P
}
/// Affine point or infinity
pub type Point = Option<(BigUint, BigUint)>;

/// Modular addition: (a + b) mod p
pub fn mod_add(a: &BigUint, b: &BigUint) -> BigUint {
    (a + b) % modp()
}

/// Modular subtraction: (a - b) mod p
pub fn mod_sub(a: &BigUint, b: &BigUint) -> BigUint {
    if a >= b {
        (a - b) % modp()
    } else {
        (modp() + a - b) % modp()
    }
}

/// Modular multiplication: (a * b) mod p
pub fn mod_mul(a: &BigUint, b: &BigUint) -> BigUint {
    (a * b) % modp()
}

/// Modular inverse: a^{-1} mod p
pub fn mod_inv(a: &BigUint) -> BigUint {
    assert!(!a.is_zero(), "modular inverse of zero is undefined");
    a.modpow(&(modp() - BigUint::from(2u32)), modp())
}

/// secp256k1 affine addition
pub fn affine_add(a: &Point, b: &Point) -> Point {
    match (a, b) {
        (None, None) => None,
        (None, Some(q)) => Some(q.clone()),
        (Some(p), None) => Some(p.clone()),
        (Some((x1, y1)), Some((x2, y2))) => {
            if x1 == x2 && mod_add(y1, y2).is_zero() {
                return None;
            }

            let lam = if x1 != x2 {
                let num = mod_sub(y2, y1);
                let den = mod_sub(x2, x1);
                mod_mul(&num, &mod_inv(&den))
            } else {
                let num = mod_mul(&BigUint::from(3u32), &mod_mul(x1, x1));
                let den = mod_mul(&BigUint::from(2u32), y1);
                mod_mul(&num, &mod_inv(&den))
            };

            let xr = mod_sub(&mod_sub(&mod_mul(&lam, &lam), x1), x2);
            let yr = mod_sub(&mod_mul(&lam, &mod_sub(x1, &xr)), y1);

            Some((xr, yr))
        }
    }
}
